### PR TITLE
Use conformance access paths in IRGen

### PR DESCRIPTION
**Pull request by rjmccall**
This series of patches — primarily the last — teaches IRGen to use conformance access paths to derive protocol conformances instead of a hacky recursive algorithm.

For now, we're just trusting the AST to figure out the best path, but it shouldn't be too hard to get the AST to give us multiple paths if they exist and then figure out which is best based on the type information currently in use within the function.

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please test.

File: lib/IRGen/GenArchetype.cpp [hunk 1]
```diff
@@ -370,22 +155,107 @@ llvm::Value *irgen::emitArchetypeWitnessTableRef(IRGenFunction &IGF,
...
   // Check immediately for an existing cache entry.
-  auto wtable = IGF.tryGetLocalTypeData(archetype,
-                  LocalTypeDataKind::forAbstractProtocolWitnessTable(proto));
+  // TODO: don't give this absolute precedence over other access paths.
+  auto wtable = IGF.tryGetLocalTypeData(archetype, localDataKind);
   if (wtable) return wtable;
 
-  // Otherwise, find the best path from one of the protocols directly
-  // conformed to by the protocol, then get that conformance.
+  // It can happen with class constraints that Sema will consider a
```
Comment (slavapestov) [inline lib/IRGen/GenArchetype.cpp]: Do you have an example of this? It looks like another consequence of improper modeling of superclass constraints.

File: lib/IRGen/GenArchetype.cpp [hunk 2]
```diff
@@ -370,22 +155,107 @@ llvm::Value *irgen::emitArchetypeWitnessTableRef(IRGenFunction &IGF,
...
+      return emitWitnessTableRef(IGF, archetype, *conformance);
+    }
+  }
+
+  // If we don't have an environment, this must be an implied witness table
+  // reference.
+  // FIXME: eliminate this path when opened types have generic environments.
+  auto environment = archetype->getGenericEnvironment();
+  if (!environment) {
+    SmallVector<ProtocolEntry, 4> entries;
```
Comment (slavapestov) [inline lib/IRGen/GenArchetype.cpp]: Maybe add an assert(archetype->isOpenedExistential()) here?

File: test/IRGen/class_bounded_generics.swift [hunk 3]
```diff
@@ -88,9 +88,6 @@ func class_bounded_archetype_method<T : ClassBoundBinary>(_ x: T, y: T) {
   // CHECK: [[INHERITED:%.*]] = load i8*, i8** %T.ClassBoundBinary, align 8
   // CHECK: [[INHERITED_WTBL:%.*]] = bitcast i8* [[INHERITED]] to i8**
   // CHECK: [[WITNESS:%.*]] = load i8*, i8** [[INHERITED_WTBL]], align 8
-  // FIXME: Redundant load of inherited conformance
```
Comment (slavapestov) [inline test/IRGen/class_bounded_generics.swift]: Sweet :)

File: lib/IRGen/GenArchetype.cpp (hunk 1 above)
Comment (rjmccall) [inline lib/IRGen/GenArchetype.cpp]: I absolutely agree.  That was existing code that I didn't feel like testing whether it was still necessary.

**Comment by rjmccall** [comment]
@swift-ci Please test.

File: lib/IRGen/GenArchetype.cpp (hunk 2 above)
Comment (rjmccall) [inline lib/IRGen/GenArchetype.cpp]: Yeah, good idea.

File: lib/IRGen/GenArchetype.cpp (hunk 1 above)
Comment (DougGregor) [inline lib/IRGen/GenArchetype.cpp]: The `GenericSignatureBuilder` should no longer create such concrete constraints; I fixed something in this area a week or so ago when teaching it to maintain all requirement sources for superclass constraints.

**Comment by DougGregor** [comment]
Fantastic! The AST doesn't currently track all of the requirement sources for conformances (which would be needed to enumerate all possible paths and pick the best one), but it does some basic keep-the-shortest-path logic.

**Pull request merged**